### PR TITLE
feat: constant and maybe types

### PR DESCRIPTION
**Pull request by mattiamanzati**
Builds on top of #35

### CODE REVIEW

File: src/types/constant.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+export class Constant extends PrimitiveType {
+    readonly value: any
+
+    constructor(value: any) {
+        super("" + value)
+        this.value = value
+    }
+
+    describe(){
+        return this.value
```
Comment (mweststrate) [inline src/types/constant.ts]: return this.name? (stringified)

File: src/types/constant.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    }
+
+    is(value) {
+        return value === this.value && super.is(value)
```
Comment (mweststrate) [inline src/types/constant.ts]: what does the `super.is` add?

File: src/types/constant.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    }
+
+}
+
+export function createConstantFactory(value: any): any {
+    invariant(isPrimitive(value), 'Constant types can be built only on top of primitives')
```
Comment (mweststrate) [inline src/types/constant.ts]: Future work could be to support (plain) complex types and deep Freeze?

File: src/types/maybe.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import {createUnionFactory} from "./union"
+import {createConstantFactory} from "./constant"
+import {createDefaultValueFactory} from "./with-default"
+import {isFactory, IFactory} from "../core/factories"
+
+export function createMaybeFactory(type: IFactory<any, any>): IFactory<any, any>{
+    return createDefaultValueFactory(createUnionFactory(createConstantFactory(null), type), null)
```
Comment (mweststrate) [inline src/types/maybe.ts]: Singleton for null factory?

File: src/types/constant.ts (hunk 3 above)
Comment (mattiamanzati) [inline src/types/constant.ts]: I inteded to do it in a separate type named "forzen" or "serializable", the problem with complex object is if the equality check should be done on reference or deep equals.

File: src/types/constant.ts (hunk 2 above)
Comment (mattiamanzati) [inline src/types/constant.ts]: Just the isPrimitive check, we can remove the extends Primitive if you want

File: src/types/constant.ts (hunk 1 above)
Comment (mattiamanzati) [inline src/types/constant.ts]: This is done to make typescript-like describes. TypeScript itself shows "a" | "b" | "c", so maybe it should be `JSON.stringify(this.value)`

File: src/types/maybe.ts (hunk 4 above)
Comment (mattiamanzati) [inline src/types/maybe.ts]: Makes sense!

**Comment by mweststrate** [comment]
Great!

**Pull request merged**